### PR TITLE
Proxy graceful termination

### PR DESCRIPTION
**Pull request by enarha**
Handle graceful termination when SIGTERM/SIGINT signal is received by the main process.

### CODE REVIEW

**Comment by enarha** [comment]
@gabemontero PTAL. The PR is marked as WIP because I'm considering adding tests around that, but some early feedback is welcome.

File: pkg/server/server.go [hunk 1]
```diff
@@ -71,28 +73,45 @@ func NewServer(host string, port int, insecureSkipTLS, insecureSkipWebhookVerify
...
+		}
+	}()
+	<-stopCh
+	zapLogger.Info("Shutting down sprayproxy")
+	// ensure graceful shutdown
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	if err := srv.Shutdown(ctx); err != nil {
+		zapLogger.Error(fmt.Sprintf("Shutdown sprayproxy error %v", err))
+	}
```
Comment (gabemontero) [inline pkg/server/server.go:112]: per what is at https://gin-gonic.com/docs/examples/graceful-restart-or-stop/ I believe you need to add
```
	// catching ctx.Done(). timeout of 5 seconds.
	select {
	case <-ctx.Done():
		zapLogger.Info("timeout of 5 seconds.")
	}
	zapLogger.Info("Server exiting")
```

However, per Adam's https://github.com/redhat-appstudio/sprayproxy/pull/20/files#r1097906460 and https://github.com/redhat-appstudio/sprayproxy/pull/20/files#r1099059894 shouldn't we be getting a logger in method here using the example he posted to that we can `ensure the logger is synced before termination` per his comment in https://issues.redhat.com/browse/PLNSRVCE-1081

Comment (enarha) [reply pkg/server/server.go:112]: Re missing block as per the gin-gonic example, I do not think it's needed. I see it's only used to log a message indicating if the server was shutdown after the timeout period was exceeded. I tested manually adding a 5 secs sleep interval in the handler:
```
^C{"level":"info","ts":"2023-10-17T14:48:35Z","caller":"server/server.go:103","msg":"Shutting down sprayproxy","audit":"true","service":"sprayproxy"}
{"level":"info","ts":"2023-10-17T14:48:38Z","caller":"proxy/proxy.go:198","msg":"proxied request","audit":"true","service":"sprayproxy","method":"POST","path":"/","query":"","insecure-tls":true,"insecure-webhook":true,"request-id":"382a39c2-30cf-44aa-9e42-4a6c9c620cfa","backend":"localhost:8081","latency":0.003926494,"status":200}
{"level":"info","ts":"2023-10-17T14:48:38Z","caller":"zap/zap.go:90","msg":"/","audit":"true","service":"sprayproxy","status":200,"method":"POST","path":"/","query":"","ip":"127.0.0.1","user-agent":"curl/8.0.1","latency":5.005257243,"request-id":"382a39c2-30cf-44aa-9e42-4a6c9c620cfa"}
``` 
and you can see the SIGTERM was sent in `14:48:35` and the request handling finished in `14:48:38` (it took 1.something secs to manage the flow manually, issue the request, jump a window and CTRL+c).
I also took the log sync in account and again the log excerpt confirms it.

Comment (gabemontero) [reply pkg/server/server.go:112]: it is best practice sort of thing that we use in a lot on controller / cli type implemention 

excluding the "timeout" scenario you looked into I'll copy/paste the godoc for `Done` and the other potential scenarios

```
	// Done returns a channel that's closed when work done on behalf of this
	// context should be canceled. Done may return nil if this context can
	// never be canceled. Successive calls to Done return the same value.
	// The close of the Done channel may happen asynchronously,
	// after the cancel function returns.
	//
	// WithCancel arranges for Done to be closed when cancel is called;
	// WithDeadline arranges for Done to be closed when the deadline
	// expires; WithTimeout arranges for Done to be closed when the timeout
	// elapses.
	//
	// Done is provided for use in select statements:
	//
	//  // Stream generates values with DoSomething and sends them to out
	//  // until DoSomething returns an error or ctx.Done is closed.
	//  func Stream(ctx context.Context, out chan<- Value) error {
	//  	for {
	//  		v, err := DoSomething(ctx)
	//  		if err != nil {
	//  			return err
	//  		}
	//  		select {
	//  		case <-ctx.Done():
	//  			return ctx.Err()
	//  		case out <- v:
	//  		}
	//  	}
	//  }
	//
	// See https://blog.golang.org/pipelines for more examples of how to use
	// a Done channel for cancellation.
```

You should not overthink this / get too clever with this IMO.  But yeah, rather than assuming it is a timeout like gin-gonic did, we should check `ctx.Err()` and log it if something unexpected has happened.

Comment (enarha) [reply pkg/server/server.go:112]: I spent more time playing with that and here are my findings. `http.Server.Shutdown(ctx)` does not cancel the context when it's done handling the requests before the actual timeout.  Using `select` and `ctx.Done()` always blocks until the context is cancelled, which means that even when the server is idle and can shutdown immediately, we still wait for the context to timeout which will make the pod restart slower.
Re checking the context error (`ctx.Err()`), I tried adding a check and log it when it's not nil. I found the same error is also returned by `http.Server.Shutdown(ctx)`, so basically it logs the same thing twice:
```
^C{"level":"info","ts":"2023-10-18T10:06:40Z","caller":"server/server.go:103","msg":"Shutting down sprayproxy","audit":"true","service":"sprayproxy"}
{"level":"error","ts":"2023-10-18T10:06:45Z","caller":"server/server.go:108","msg":"Shutdown sprayproxy error: context deadline exceeded","audit":"true","service":"sprayproxy"}
{"level":"error","ts":"2023-10-18T10:06:45Z","caller":"server/server.go:112","msg":"Graceful shutdown error: context deadline exceeded","audit":"true","service":"sprayproxy"}
```
Bottom line is, I do not really see any value in changing that part and it adds some delay when we wait for `ctx.Done()`. If you still think we should add that, then OK, it's an easy change I already tested.

Comment (gabemontero) [reply pkg/server/server.go:112]: Interesting.

Put a comment in the code with your findings, and hence why we are diverting from the example provided in the `gin-gonic` doc and not doing the select / case <- ctx.Done(), and we'll go with what you have.

Then in the future, if a scenario in the future arises that somehow is outside your investigations / testings, we'll have some context.

thanks

**Review by gabemontero — APPROVED** [review]
sort out the codecov /  unit test failure, remove the WIP designation in the PR title, and I'm OK with merging this @enarha
